go/parquet/metadata/column_chunk.go

```diff
@@ -0,0 +1,385 @@
+// and it's associated Column
+type ColumnChunkMetaData struct {
+	column        *format.ColumnChunk
+	columnMeta    *format.ColumnMetaData
+	decryptedMeta format.ColumnMetaData
+	descr         *schema.Column
+	writerVersion *AppVersion
+	encodings     []parquet.Encoding
+	encodingStats []format.PageEncodingStats
+	possibleStats TypedStatistics
+	mem           memory.Allocator
+}
+
+// NewColumnChunkMetaData creates an instance of the metadata from a column chunk and descriptor
+//
+// this is primarily used internally or between the subpackages. ColumnChunkMetaDataBuilder should
+// be used by consumers instead of using this directly.
+func NewColumnChunkMetaData(column *format.ColumnChunk, descr *schema.Column, writerVersion *AppVersion, rowGroupOrdinal, columnOrdinal int16, fileDecryptor encryption.FileDecryptor) (*ColumnChunkMetaData, error) {
+	c := &ColumnChunkMetaData{
+		column:        column,
+		columnMeta:    column.GetMetaData(),
+		descr:         descr,
+		writerVersion: writerVersion,
+		mem:           memory.DefaultAllocator,
+	}
+	if column.IsSetCryptoMetadata() {
+		ccmd := column.CryptoMetadata
+
+		if ccmd.IsSetENCRYPTION_WITH_COLUMN_KEY() {
+			if fileDecryptor != nil && fileDecryptor.Properties() != nil {
+				// should decrypt metadata
+				path := parquet.ColumnPath(ccmd.ENCRYPTION_WITH_COLUMN_KEY.GetPathInSchema())
+				keyMetadata := ccmd.ENCRYPTION_WITH_COLUMN_KEY.GetKeyMetadata()
+				aadColumnMetadata := encryption.CreateModuleAad(fileDecryptor.FileAad(), encryption.ColumnMetaModule, rowGroupOrdinal, columnOrdinal, -1)
+				decryptor := fileDecryptor.GetColumnMetaDecryptor(path.String(), string(keyMetadata), aadColumnMetadata)
+				thrift.DeserializeThrift(&c.decryptedMeta, decryptor.Decrypt(column.GetEncryptedColumnMetadata()))
+				c.columnMeta = &c.decryptedMeta
+			} else {
+				return nil, xerrors.New("cannot decrypt column metadata. file decryption not setup correctly")
+			}
+		}
+	}
+	for _, enc := range c.columnMeta.Encodings {
+		c.encodings = append(c.encodings, parquet.Encoding(enc))
+	}
+	for _, enc := range c.columnMeta.EncodingStats {
+		c.encodingStats = append(c.encodingStats, *enc)
+	}
+	return c, nil
+}
+
+// CryptoMetadata returns the cryptographic metadata for how this column was
+// encrypted and how to decrypt it.
+func (c *ColumnChunkMetaData) CryptoMetadata() *format.ColumnCryptoMetaData {
+	return c.column.GetCryptoMetadata()
+}
+
+// FileOffset is the location in the file where the column data begins
+func (c *ColumnChunkMetaData) FileOffset() int64 { return c.column.FileOffset }
+
+// FilePath gives the name of the parquet file if provided in the metadata
+func (c *ColumnChunkMetaData) FilePath() string { return c.column.GetFilePath() }
+
+// Type is the physical storage type used in the parquet file for this column chunk.
+func (c *ColumnChunkMetaData) Type() parquet.Type { return parquet.Type(c.columnMeta.Type) }
+
+// NumValues is the number of values stored in just this chunk
```

Comment:
can we be more specific.  Does this number include nulls?

Reply:
updated comment.